File: IT/Oggetti/Esercizi/Python/esercizio16.py
# file esercizio16.py

from __future__ import print_function

class Engine(object):
	def __init__(self, cc):
		self.__cc = cc

	def __str__(self):
		return "motore di {} centimetri cubi".format(self.__cc)

class Car(object):
	def __init__(self, model, vin, plate, engine):
		self.model = model
		self.vin = vin
		self.plate = plate
		self.engine = engine

	def getModel(self):
		return self.__model

	def setModel(self, model):
		self.__model = model

	def getVin(self):
		return self.__vin

	def setVin(self, vin):
		self.__vin = vin

	def getPlate(self):
		return self.__plate

	def setPlate(self, plate):
		self.__plate = plate

	def getEngine(self):
		return self.__engine

	def setEngine(self, engine):
		self.__engine = engine

	model = property(getModel, setModel)
	vin = property(getVin, setVin)
	plate = property(getPlate, setPlate)
	engine = property(getEngine, setEngine)

	def __str__(self):
		return "{}, targa: {}, n.telaio: {}, {}.".format(self.model, self.plate, self.vin, self.engine)

fiat500 = Car(model="Fiat 500", vin="1G8AN12F04Z242828", plate="AS 123 DF", engine=Engine(cc=1400))
classeA = Car(model="Mercedes Classe A", vin="2FZHAZBD21AH02482", plate="CA 128 TD", engine=Engine(cc=1600))
alfaGiulia = Car(model="Alfa Romeo Giulia", vin="2G1115SL7F9174905", plate="EE 119 CA", engine=Engine(cc=2000))

print(fiat500)
print(classeA)
print(alfaGiulia)

File: IT/Oggetti/Esercizi/C++/esercizio16.cpp
/* file esercizio16.cpp */

#include <iostream>

using std::cout;
using std::endl;
using std::string;
using std::ostream;

class Engine {
	private:
		int cc;

	public:
		Engine(int);

	friend ostream& operator<<(ostream&, const Engine&);
};

Engine::Engine(int cc) {
	this->cc = cc;
}

class Car {
	private:
		string model;
		string plate;
		string vin;
		Engine* engine;

	public:
		Car(string, string, string, Engine*);
		~Car();

		void setModel(string);
		string getModel();
		void setPlate(string);
		string getPlate();
		void setVin(string);
		string getVin();
		Engine* getEngine();
		void setEngine(Engine*);

	friend ostream& operator<<(ostream&, const Car&);
};

ostream& operator<<(ostream &os, const Engine& engine) {
	os << "motore di " << engine.cc << " centimetri cubi";
	return os;
}

Car::Car(string model, string plate, string vin, Engine* engine) {
	setModel(model);
	setPlate(plate);
	setVin(vin);
	setEngine(engine);
}

Car::~Car() {
	delete engine;
}

string Car::getModel() {
	return this->model;
}

void Car::setModel(string model) {
	this->model = model;
}

string Car::getPlate() {
	return this->plate;
}

void Car::setPlate(string plate) {
	this->plate = plate;
}

string Car::getVin() {
	return this->vin;
}

void Car::setVin(string vin) {
	this->vin = vin;
}

Engine* Car::getEngine() {
	return this->engine;
}

void Car::setEngine(Engine* engine) {
	this->engine = engine;
}

ostream& operator<<(ostream &os, const Car& car) {
	os << car.model << ", targa: " << car.plate << ", n. telaio: " << car.vin << ", " << *car.engine << ".";
	return os;
}

int main()
{
	Car fiat500("Fiat 500", "AS 123 DF", "1G8AN12F04Z242828", new Engine(1400));
	Car classeA("Mercedes Classe A", "CA 128 TD", "2FZHAZBD21AH02482", new Engine(1600));
	Car alfaGiulia("Alfa Romeo Giulia", "EE 119 CA", "2G1115SL7F9174905", new Engine(2000));

	cout << fiat500 << endl;
	cout << classeA << endl;
	cout << alfaGiulia << endl;
}

